### ISSUE TITLE
Automatic namespace resolution for custom exceptions

### ISSUE THREAD

**Issue by caseyamcl** [Respect/Validation#1278]
I noticed that in version 2, you need to manually add the exception namespace to the factory class as follows:

```php
$factory = (new \Respect\Validation\Factory())->withExceptionNamespace("TestRespect\\Exceptions");
\Respect\Validation\Factory::setDefaultInstance($factory);
```

Otherwise, you'll get the default `ValidationException` every time, regardless of whether or not you create your own custom exceptions?

This involves a lot of boiler-plate code for my use-case, and it would be easily remediated by using the `ReflectionClass`, which appears to already be initialized in the Factory class.

Am I doing something wrong, or is this the expected behavior?  If not, then I'd be happy to submit a PR to correct this.

**Comment by henriquemoody** [comment]
Hi, @caseyamcl! 

Yes, indeed you need to manually add the exception namespace to the factory, or else you indeed get `ValidationException` every time. I made that choice for two reasons:

1) **Not all rules need a specific exception**: that gives users more flexibility when using this library. Before you needed to create an exception, even if it was just a placeholder.
2) **It makes explicit what's the exception namespace**: that gives users more flexibility when choosing when they want to put their exceptions. If this library didn't have the many rules they have, I would particularly prefer to keep the exceptions next to the rules.

Can you elaborate a bit better how this involves a lot more of boiler-plate in your use-case? I find it a bit hard to understand because the only effort I see is to add another line when defining the default factory, taking your code as an example, it would be:

```php
use Respect\Validation\Factory;
    
Factory::setDefaultInstance(
    (new Factory())
        ->withRuleNamespace('TestRespect\\Rules')
        ->withExceptionNamespace('TestRespect\\Exceptions')
);
```

**Comment by caseyamcl** [comment]
Thanks for the clarification, @henriquemoody.

I read [the documentation](https://respect-validation.readthedocs.io/en/2.0/custom-rules/) several times, and it wasn't clear to me that this behavior had changed, so I didn't know if it was intentional or not.

The overhead that I believe is added versus v1 is the dependency on a service class, so you can't just run `Validator::[rule]` anymore, unless you set the default factory, but that comes with its own complication.

Whereas before you could just run `Validator::[rule]` and expect custom exceptions to "just work", now you've got to inject the `Factory` service class.

For example, I am working on my own experimental library, [openapi-params](https://github.com/caseyamcl/openapi-params) that I intend to implement if it is successful.  But if I want to upgrade to v2, I've got to make a decision of where to inject the `Factory` service class in the very probable case that I'll have to pass that around as a dependency to my own library; since my library is a mid-level abstraction, it is meant to be used in a framework that will likely also have custom validation rules and exceptions, so I anticipate that I'll need a single `Factory` instance no matter what.

Regardless of whether you accept this PR or not, I'll find a way to make it work.

**Comment by henriquemoody** [comment]
Thanks for your reply, @caseyamcl 

In both versions 1.x or 2.x of Validation you don't need to add the namespace of your own rules unless you want to access them via the Validator class. But in both versions you need to do that

Version 1.x

```php
Validator::with('OpenApiParams\\Validation\\Rules');
```

Version 2.x

```php
Factory::setDefaultInstance(
    (new Factory())
        ->withRuleNamespace('OpenApiParams\\Validation\\Rules')
);
```

The only difference is that you would need to specify the exception namespaces in the factory.

```
Factory::setDefaultInstance(
    (new Factory())
        ->withRuleNamespace('OpenApiParams\\Validation\\Rules')
        ->withRuleNamespace('OpenApiParams\\Validation\\Exceptions')
);
```

I do understand, though, that before, the exception was creating using a _convention_ based on the convention of the namespaces of the rule, so, if you want to use a rule, directly, like `new ValidUnixPath()`, that rule will indeed not throw an exception based on that namespace.

I'm still not quite sure how to fix that in a nice way, because I don't wanna use conventions anymore, until then, I suggest you to use a custom autoloader for composer:

```json
  "autoload": {
    "psr-4": {
      "OpenApiParams\\": "src",
    },
    "files": ["bootstrap.php"]
  },
```

Then doing the magic in "bootstrap.php"

**Comment by caseyamcl** [comment]
Hi @henriquemoody, I've found a workaround, but it would be much easier with the following methods available:

```
(new Factory())->hasRuleNamespace(string $namespace): bool;
(new Factory())->hasExceptionNamespace(string $namespace): bool;
```

Basically, it allows you to check if a namespace is registered or not, which would be useful for my use-case while preserving immutability.

What do you think about me closing this pull request and opening another one with those methods added?

**Comment by oschwald** [comment]
I agree that the custom rules documentation is not clear on this. It seems to suggest that the convention-based namespaces is required and that overwriting the default `Factory` is only necessary if you are using the rules in a chain.

As far as the new approach of getting the exception namespace from the default `Factory`, do you have any guidance on how this is supposed to work if you are using this library in another library? Overwriting shared global state does not seem ideal. Another library or the application itself may also be setting the default `Factory`.

**Comment by henriquemoody** [comment]
This behaviour was fixed on 8b2819e9f8d43c67bd6bac676eb260ed59b01de9, and will be available on version 2.1.

Thanks for all your comments! 🐼

**Comment by oschwald** [comment]
Thank you!
